# origin: bk (linux-2.6)
# cset: 1.1371.776.19 (linux-2.6)
# inclusion: n/a (only in linux-2.6 bk)
# description: fix SMP deadlock for aic7xxx
# revision date: 2004-09-03

# This is a BitKeeper generated diff -Nru style patch.
#
# ChangeSet
#   2004/05/04 09:21:23-05:00 akpm@osdl.org 
#   [PATCH] aic7xxx deadlock fix
#   
#   We cannot call del_timer_sync() from within that timer's handler function!
# 
# drivers/scsi/aic7xxx/aic7xxx_osm.c
#   2004/04/26 01:09:29-05:00 akpm@osdl.org +9 -3
#   aic7xxx deadlock fix
# 
diff -Nru a/drivers/scsi/aic7xxx/aic7xxx_osm.c b/drivers/scsi/aic7xxx/aic7xxx_osm.c
--- a/drivers/scsi/aic7xxx/aic7xxx_osm.c	2004-09-03 20:40:42 -07:00
+++ b/drivers/scsi/aic7xxx/aic7xxx_osm.c	2004-09-03 20:40:42 -07:00
@@ -3973,11 +3973,10 @@
 }
 
 static void
-ahc_linux_free_device(struct ahc_softc *ahc, struct ahc_linux_device *dev)
+__ahc_linux_free_device(struct ahc_softc *ahc, struct ahc_linux_device *dev)
 {
 	struct ahc_linux_target *targ;
 
-	del_timer_sync(&dev->timer);
 	targ = dev->target;
 	targ->devices[dev->lun] = NULL;
 	free(dev, M_DEVBUF);
@@ -3987,6 +3986,13 @@
 		ahc_linux_free_target(ahc, targ);
 }
 
+static void
+ahc_linux_free_device(struct ahc_softc *ahc, struct ahc_linux_device *dev)
+{
+	del_timer_sync(&dev->timer);
+	__ahc_linux_free_device(ahc, dev);
+}
+
 void
 ahc_send_async(struct ahc_softc *ahc, char channel,
 	       u_int target, u_int lun, ac_code code, void *arg)
@@ -4697,7 +4703,7 @@
 		ahc_linux_run_device_queue(ahc, dev);
 	if (TAILQ_EMPTY(&dev->busyq)
 	 && dev->active == 0)
-		ahc_linux_free_device(ahc, dev);
+		__ahc_linux_free_device(ahc, dev);
 	ahc_unlock(ahc, &s);
 }
 
